### PR TITLE
[WIP] Refactor comparison

### PR DESCRIPTION
**Pull request by bookwar**
Complete rewrite of the eln-check script

* Create BuildSource object which might point to Koji tag or Brew tag
* Create Comparison object, which stores data about comparison results
* Render html, json and txt output via templates engine
* Allow full comparison or comparison per package

### CODE REVIEW

File: compare.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+            return None
+
+    def make_cache(self):
+        """Fetch all builds from tag"""
+
+        logging.debug(f'Make cache for {self}...')
+        builds = self.infra.listTagged(self.tag, latest=True)
+        for build in builds:
+            self.cache[build["name"]] = build
+        logging.debug(f'Done Make cache for {self}')
```
Comment (jimbair) [inline compare.py]: Small nitpick: I might reword this to "Done making cache".

File: compare.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+    def __init__(self, content, source1, source2):
+        self.content = content
+        self.source1 = source1
+        self.source2 = source2
+
+        self.results = {}
+
+    def compare_one(self, package):
+        """Returm comparison data for a package
```
Comment (jimbair) [inline compare.py]: typo :)

**Comment by jimbair** [comment]
Overall, it looks good to me; just a few suggestions listed above. :)

File: compare.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        if source_id == "rawhide":
+            infra = koji.ClientSession('https://koji.fedoraproject.org/kojihub')
+            tag = infra.getFullInheritance('rawhide')[0]['name']
+        if source_id == "eln":
+            infra = koji.ClientSession('https://koji.fedoraproject.org/kojihub')
+            tag = "eln"
+        if source_id == "rhel":
+            # FIXME
+            infra = koji.ClientSession('https://brew')
+            tag = "rhel-9.0.0-alpha"
```
Comment (sturivny) [inline compare.py:47]: I'd also added a condition here where `source_id ` doesn't equal all these cases.
It will help to debug if something goes wrong.

File: compare.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+
+def get_content(distro_view="eln"):
+    """Builds the full list of packages for the distro from the Content Resolver
+
+    Merges result for all architectures.
+    """
+    merged_packages = set()
+
+    distro_url = "https://tiny.distro.builders"
```
Comment (sturivny) [inline compare.py:217]: All those constant should be in a separate place.

File: compare.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+    source1 = BuildSource(source_id=args.source1, make_cache=args.cache)
+    source2 = BuildSource(source_id=args.source2, make_cache=args.cache)
+
+    if args.packages:
+        content = args.packages
+    else:
+        content = sorted(get_content())
+
+    C = Comparison(content, source1, source2)
```
Comment (sturivny) [inline compare.py:333]: Not the best name for the instance :)

File: compare.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    """Get epoch, version, release data from the build
+
+    We currently reset Epoch value to 0, because we have number of cases where
+    epoch of a package in Rawhide is different from that of ELN.
+
+    We remove dist tag data from the release, so that we can compare nvr's
+    between different distributions.
+    """
+
+    epoch = "0"
```
Comment (sturivny) [inline compare.py:249]: Is it necessary to create a separate variable here?
Couldn't find where it uses instead of `return` .

File: compare.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    for arch in arches:
+        for this_source in which_source:
+            url = (
+                "{distro_url}"
+                "/view-{this_source}-package-name-list--view-{distro_view}--{arch}.txt"
+            ).format(distro_url=distro_url, this_source=this_source, distro_view=distro_view, arch=arch)
+
+            logging.debug("downloading {url}".format(url=url))
+
+            r = requests.get(url, allow_redirects=True)
```
Comment (sturivny) [inline compare.py:230]: Please, use `try` construction for requests.

File: templates/status.json.j2 [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+"date": "{{ date }}",
+"source1": "{{ source1 }}",
+"source2": "{{ source2 }}",
+"stats": {{ stats|tojson }},
+"packages": {{ results|tojson }}
+}
```
Comment (sturivny) [inline templates/status.json.j2:7]: Add a newline at the end of this file.

**Review by sturivny — COMMENTED** [review]
Would be great to add `some` tests here ;)

File: compare.py (hunk 1 above)
Comment (bookwar) [reply compare.py]: Fixed

File: compare.py (hunk 2 above)
Comment (bookwar) [reply compare.py]: Fixed, thanks
